app/views/users/confirmations/_invalid_token.html.haml

```diff
@@ -0,0 +1,28 @@
+- content_for(:title, 'Confirmez votre adresse email')
+
+.container.devise-container.devise-confirmations
+  .one-column-centered
+    %h1.center Votre jeton de connection est invalide
+
+    %p.confirmation-preamble
+      Lors de la création de compte, un jeton aléatoire est généré afin d'identifier la demande.
+      Ce jeton nous est ensuite transmis à nouveau lorsque vous cliquez sur l'email d'activation : cela permet de garantir
+      que vous avez bien demandé la création d'un compte, puis que vous possédez bien l'adresse email utilisée.
+
+    %p.confirmation-preamble
+      Le jeton que vous venez d'utiliser est invalide : cela peut arriver
+      s'il est trop vieux ou lors une erreur de copier/coller, par exemple.
+      Nous ne pouvons donc pas activer votre compte, car nous n'avons pas de garantie que vous avez bien fait la demande initiale.
+
+    %p.confirmation-instructions
+      %strong Afin d'activer votre compte,
+      vous devez utiliser un autre jeton de confirmation.
+      Nous pouvons vous en renvoyer un avec le formulaire ci-dessous.
+
+    .confirmation-resend
+
+      = form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { class: 'form' }) do |f|
+        = f.label :email, 'Votre email'
```

Comment:
Tu crois que ça serait simple de pré-remplir ce champ avec l'email associé au jeton ? Ou c'est galère ?

Reply:
J'y ai pensé, et ce serait mieux niveau UX ; même si des fois on a pas l'email (ex: token tronqué, donc incorrect et non expiré), ça permet d'accélerer le process dans plein de cas. Surtout que souvent, on oublie quel email on a utilisé, et on se retrouve avec un 2nd compte sur une nouvelle adresse…

Le problème par contre, c'est que ça ouvre la porte à l'énumération d'utilisateurs, donc ça introduit une faille de sécurité. Aujourd'hui c'est capillotracté (l'espace de recherche est immense), mais demain si une attaque réaliste est découverte dans `devise`, on est exposé.

Reply:
Je me suis un poil arraché les cheveux pour comprendre pourquoi je n'arrivais pas à afficher l'email : en théorie, on peut exploiter `resource.email` et afficher l'email. En pratique, nos comptes n'ont pas de durée limite de confirmation ([`config.confirm_within`](https://www.rubydoc.info/github/heartcombo/devise/master/Devise/Models/Confirmable) est commenté dans `initializers/devise.rb`), donc:
 - les comptes qui accèdent depuis un vieux lien seront donc confirmés automatiquement, et sont redirigés sur l'écran de sign-in
 - cet écran ne s'affichera que pour les utilisateurs qui ont copié un lien invalide (erreur de copier/coller, par ex) et pour lesquels on ne trouve pas de lien de confirmation.

Je me dis qu'on peut laisser la conf en l'état.